libfreerdp/codec/h264_mediacodec.c

```diff
@@ -0,0 +1,516 @@
+typedef media_status_t (*AMediaCodec_getName_t)(AMediaCodec*, char**); // 28
+typedef void (*AMediaCodec_releaseName_t)(AMediaCodec*, char*); // 28
+typedef AMediaFormat* (*AMediaCodec_getInputFormat_t)(AMediaCodec*); // 28
+
+const int COLOR_FormatYUV420Planar = 19;
+const int COLOR_FormatYUV420Flexible = 0x7f420888;
+
+struct _H264_CONTEXT_MEDIACODEC
+{
+    AMediaCodec* decoder;
+    AMediaFormat* inputFormat;
+    AMediaFormat* outputFormat;
+    int32_t width;
+    int32_t height;
+    ssize_t currentOutputBufferIndex;
+
+    // libmediandk.so imports
+    HMODULE mediandkLibrary;
+
+    AMediaFormat_new_t fnAMediaFormat_new;
+    AMediaFormat_delete_t fnAMediaFormat_delete;
+    AMediaFormat_toString_t fnAMediaFormat_toString;
+    AMediaFormat_setInt32_t fnAMediaFormat_setInt32;
+    AMediaFormat_setString_t fnAMediaFormat_setString;
+    AMediaCodec_createDecoderByType_t fnAMediaCodec_createDecoderByType;
+    AMediaCodec_delete_t fnAMediaCodec_delete;
+    AMediaCodec_configure_t fnAMediaCodec_configure;
+    AMediaCodec_start_t fnAMediaCodec_start;
+    AMediaCodec_stop_t fnAMediaCodec_stop;
+    AMediaCodec_getInputBuffer_t fnAMediaCodec_getInputBuffer;
+    AMediaCodec_getOutputBuffer_t fnAMediaCodec_getOutputBuffer;
+    AMediaCodec_dequeueInputBuffer_t fnAMediaCodec_dequeueInputBuffer;
+    AMediaCodec_queueInputBuffer_t fnAMediaCodec_queueInputBuffer;
+    AMediaCodec_dequeueOutputBuffer_t fnAMediaCodec_dequeueOutputBuffer;
+    AMediaCodec_getOutputFormat_t fnAMediaCodec_getOutputFormat;
+    AMediaCodec_releaseOutputBuffer_t fnAMediaCodec_releaseOutputBuffer;
+    AMediaCodec_setParameters_t fnAMediaCodec_setParameters;
+    AMediaCodec_getName_t fnAMediaCodec_getName;
+    AMediaCodec_releaseName_t fnAMediaCodec_releaseName;
+    AMediaCodec_getInputFormat_t fnAMediaCodec_getInputFormat;
+
+    const char* gAMediaFormatKeyMime;
+    const char* gAMediaFormatKeyWidth;
+    const char* gAMediaFormatKeyHeight;
+    const char* gAMediaFormatKeyColorFormat;
+};
+
+typedef struct _H264_CONTEXT_MEDIACODEC H264_CONTEXT_MEDIACODEC;
+
+static int load_libmediandk(H264_CONTEXT* h264)
+{
+    H264_CONTEXT_MEDIACODEC* sys = (H264_CONTEXT_MEDIACODEC*)h264->pSystemData;
+    WLog_Print(h264->log, WLOG_INFO, "MediaCodec Loading libmediandk.so");
+
+    sys->mediandkLibrary = LoadLibraryA("libmediandk.so");
+    if (sys->mediandkLibrary == NULL) 
+    {
+        WLog_Print(h264->log, WLOG_ERROR, "Error loading libmediandk.so");
+        return -1;
+    }
+
+#define RESOLVE_MEDIANDK_FUNC(name) \
+    sys->fn##name = GetProcAddress(sys->mediandkLibrary, #name); \
+    if (sys->fn##name == NULL) \
+    { \
+        WLog_Print(h264->log, WLOG_ERROR, "Error resolving function " #name " from libmediandk.so"); \
+        return -1; \
+    }
+
+    const char **temp = NULL;
+#define RESOLVE_MEDIANDK_VARIABLE(member, exported) \
+    temp = GetProcAddress(sys->mediandkLibrary, exported); \
```

Comment:
Maybe do a scope (e.g. `{ const char** temp = GetProcAddress... }` to avoid the variable outside the macro.
Also, same as above, might add `sys` to the macro arguments.